jwst/outlier_detection/outlier_detection.py

```diff
@@ -271,18 +243,20 @@ def create_median(self, resampled_models):
                     elem1[elem2] = np.nan
 
             del badmasks
-
-            # For a stack of images with "bad" data replaced with Nan
-            # use np.nanmedian to compute the median.
-            with warnings.catch_warnings():
-                warnings.filterwarnings(action="ignore",
-                                        message="All-NaN slice encountered",
-                                        category=RuntimeWarning)
-                median_image[row1:row2] = np.nanmedian(resampled_sci, axis=0)
+            
+            # compute median using selected algorithm
+            median_computer = algorithm()
+            median_computer.w = self.outlierpars.get("rolling_window", 25)
+            med = median_computer.compute_median(resampled_sci)
+            if i ==0:
+                median_image = med
+            else:
+                median_image = np.concatenate((median_image, med), axis=1)
```

Comment:
What would be required to keep the pre-allocation of `median` here instead of using `concatenate`? Every call to concatenate will allocate a new array (for the new `median_array`) and copy the contents of the old `median_array` into the new array. This results in a ~2x memory footprint and a copy for each section (which is ~1MB of each image).

Reply:
Yeah, I think the second way is the most memory-efficient, especially if concatenation is done with something other than np.concatenate, as in [this stackoverflow answer](https://stackoverflow.com/a/51399815).

Reply:
Would you help me understand how the old and new shapes differed? I'm not familiar with tso data. Do the data arrays differ in shape?

Reply:
We used to be computing the simple median over the entire time axis, and now we need the rolling median, which will change with time.  The output array from `median_computer.compute_median()` is now 3-D instead of 2-D, and so is the output `median_image` from the whole `create_median` function.

Reply:
the dimensions of `resampled_sci` are (time, x, y)

Reply:
Thanks! Can this update be made in a way that doesn't change what's done for non-tso imaging data? I see you've added `outlier_detection_tso.py` to handle the `tso` data. Could the modified median calculation be added there so that for non-tso data the median can be pre-allocated?

Reply:
yes, that could be done. It would require repeating the majority of create_median, although perhaps there's a different refactor that could avoid this.

Reply:
I'd have to look in more detail but the majority of the complication in `create_median` is an attempt to avoid having all models in memory at the same time (the main goal of the `get_sections` API). Is that a possibility for `tso` data? If not (if prior code in this step or in the pipeline already loads all the models) it is likely that a simpler (and more efficient) median calculation could be done without the `get_sections` API.

Reply:
It would be possible to keep the rolling medians the same size in memory as the `resampled_sections`, and they could be stored in a ModelContainer of the same shape, if we didn't want/need to compute the median across sections.

Can you tell me how `get_sections()` decides where the section cutoffs are?

It seems that `get_sections()` partitions data in the spatial dimension, and the time dimension remains intact - is that correct?  If so, then what you're suggesting is not that hard to do.

Reply:
The dimensions of the sections are calculated during `ModelContainer.set_buffer`:
https://github.com/spacetelescope/jwst/blob/8b254ae2b69fed42c413f4ff7163ff2f983dfa59/jwst/datamodels/container.py#L576
It picks a number of rows that occupy 1 MB of memory (for each model, rounded up).
I think you're correct about the partitioning. If `time` is the across-model dimension for the container than the sections will contain a selection of rows for all timepoints.
The `get_sections` API is very inefficient for open models as it reopens (clones) every model in the contain for every section.